doc/man/getlist.1
.\" Automatically generated by Pod::Man 4.07 (Pod::Simple 3.32)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GETLIST 1"
.TH GETLIST 1 "2015-09-12" "INN 2.6.3" "InterNetNews Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
getlist \- Get a list from an NNTP server
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBgetlist\fR [\fB\-AR\fR] [\fB\-h\fR \fIhost\fR] [\fB\-p\fR \fIport\fR]
[\fIlist\fR [\fIpattern\fR [\fItypes\fR]]]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBgetlist\fR obtains a list from an \s-1NNTP\s0 server and sends it to standard
output.  \fIlist\fR may be one of \f(CW\*(C`active\*(C'\fR (which is the default value),
\&\f(CW\*(C`active.times\*(C'\fR, \f(CW\*(C`distrib.pats\*(C'\fR, or \f(CW\*(C`newsgroups\*(C'\fR.  \f(CW\*(C`active\*(C'\fR contains
a list of all newsgroups carried by the server with the high and low article
numbers in the group, and the group status field.  \f(CW\*(C`active.times\*(C'\fR is a list of
newsgroups, their creation times (in seconds since epoch), and the creator
of the group.  \f(CW\*(C`distrib.pats\*(C'\fR is a list of relevant distributions and
the newsgroups they apply to.  \f(CW\*(C`newsgroups\*(C'\fR is a list of newsgroups
along with their short descriptions.
.PP
For more information on the formats of these files, see \fIactive\fR\|(5),
\&\fIactive.times\fR\|(5), \fIdistrib.pats\fR\|(5) and \fInewsgroups\fR\|(5).
.PP
The listing files other than the \fIactive\fR file are common extensions to the
\&\s-1NNTP\s0 protocol and may not be available on all servers.  For instance, \f(CW\*(C`counts\*(C'\fR
\&\f(CW\*(C`distributions\*(C'\fR, \f(CW\*(C`headers\*(C'\fR, \f(CW\*(C`moderators\*(C'\fR, \f(CW\*(C`motd\*(C'\fR, \f(CW\*(C`overview.fmt\*(C'\fR and
\&\f(CW\*(C`subscriptions\*(C'\fR, amongst other, may be available.  Moreover, a \fIuwildmat\fR\|(3)
pattern \fIpattern\fR may also be usable for some of these listing files.
.PP
For more information on the formats of these files, see \fIdistributions\fR\|(5),
\&\fImoderators\fR\|(5), \fImotd.news\fR\|(5) and \fIsubscriptions\fR\|(5).  The overview fields
obtained with \f(CW\*(C`overview.fmt\*(C'\fR are the ones for which the overview database
is consistent (see \fIextraoverviewadvertised\fR in \fIinn.conf\fR for more
information).  The list obtained with \f(CW\*(C`headers\*(C'\fR contains the fields
that can be retrieved using the \s-1HDR\s0 command (\fIpattern\fR can then be
either \f(CW\*(C`msgid\*(C'\fR or \f(CW\*(C`range\*(C'\fR).  The list obtained with \f(CW\*(C`counts\*(C'\fR is like
the \f(CW\*(C`active\*(C'\fR one except that the number of articles in a newsgroup
is mentioned just before the flag of this newsgroup.
.PP
The \fIpattern\fR parameter may be used with a \fIlist\fR value of \f(CW\*(C`active\*(C'\fR,
\&\f(CW\*(C`active.times\*(C'\fR or \f(CW\*(C`newsgroups\*(C'\fR to limit the output.  If given, only
entries corresponding to newsgroups that match the \fIuwildmat\fR\|(3) pattern
\&\fIpattern\fR will be printed.
.PP
If the \fIlist\fR parameter is \f(CW\*(C`active\*(C'\fR, a third parameter, \fItypes\fR, may
also be used to restrict the list.  If given, only entries corresponding
to newsgroups with a newsgroup mode found in \fItypes\fR will be printed.
For example, a \fItypes\fR value of \f(CW\*(C`y\*(C'\fR would only show unmoderated groups,
and a \fItypes\fR value of \f(CW\*(C`ym\*(C'\fR would show all unmoderated and moderated
groups but not aliased groups or groups that disallow postings.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-A\fR" 4
.IX Item "-A"
Try to authenticate using the username and password information in
\&\fIpasswd.nntp\fR\|(5) before issuing the \s-1LIST\s0 command.
.IP "\fB\-h\fR \fIhost\fR" 4
.IX Item "-h host"
Connect to the server \fIhost\fR rather than to the value of \fIserver\fR in
\&\fIinn.conf\fR or the value of the \s-1NNTPSERVER\s0 environment variable.
.IP "\fB\-p\fR \fIport\fR" 4
.IX Item "-p port"
Connect to the port \fIport\fR, which is by default \f(CW119\fR.
.IP "\fB\-R\fR" 4
.IX Item "-R"
Try to switch to \fBnnrpd\fR with a \s-1MODE READER\s0 command before issuing the
\&\s-1LIST\s0 request.  It can be useful in case a feeding access exists for
the news server:  more information can be retrieved from \fBnnrpd\fR,
compared with \fBinnd\fR.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
The following command will obtain the one-line descriptions of all
newsgroups found on \s-1UUNET:\s0
.PP
.Vb 1
\&    getlist \-h news.uu.net newsgroups
.Ve
.PP
The following command will show the active file entries for all moderated
comp.* groups on the default server:
.PP
.Vb 1
\&    getlist active \*(Aqcomp.*\*(Aq m
.Ve
.PP
The following command lists all groups where local postings are permitted,
are moderated or aliased:
.PP
.Vb 1
\&    getlist active \*(Aq*\*(Aq ym=
.Ve
.PP
Note that the wildcard character \f(CW\*(C`*\*(C'\fR has to be quoted by simple quotes
to protect it from the shell.
.SH "HISTORY"
.IX Header "HISTORY"
Written by Landon Curt Noll <chongo@toad.com> for InterNetNews.  Rewritten
in \s-1POD\s0 by Russ Allbery <eagle@eyrie.org>.
.PP
\&\f(CW$Id:\fR getlist.pod 9767 2014\-12\-07 21:13:43Z iulius $
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIactive\fR\|(5), \fIactive.times\fR\|(5), \fIdistrib.pats\fR\|(5), \fIdistributions\fR\|(5), \fIinn.conf\fR\|(5),
\&\fImoderators\fR\|(5), \fImotd.news\fR\|(5), \fInewsgroups\fR\|(5), \fInnrpd\fR\|(8), \fIpasswd.nntp\fR\|(5),
\&\fIsubscriptions\fR\|(5), \fIuwildmat\fR\|(3).


NEWS
Changes in 2.6.3

    * Fixed the selection of the elliptic curve to use with OpenSSL 1.1.0 or
      later; NIST P-256 was enforced instead of using the most secure curve.

    * A new inn.conf parameter has been added to fine-tune the cipher suites
      to use with TLS 1.3:  the *tlsciphers13* now permits configuring them.
      A separate cipher suite configuration parameter is needed for TLS 1.3
      because TLS 1.3 cipher suites are not compatible with TLS 1.2, and
      vice-versa.  In order to avoid issues where legacy TLS 1.2 cipher
      suite configuration configured in the *tlsciphers* parameter would
      inadvertently disable all TLS 1.3 cipher suites, the inn.conf
      configuration has been separated out.

    * Fixed a regression since INN 2.6.1 that prevented articles with
      internationalized header fields (that is to say encoded in UTF-8) from
      being posted.

    * Support for Python 3 has been added to INN.  Embedded Python filtering
      and authentication hooks for innd and nnrpd can now use version 3.3.0
      or later of the Python interpreter.  In the 2.x series, version 2.3.0
      or later is still supported.

      When configuring INN with the --with-python flag, the "PYTHON"
      environment variable, when set, is used to select the interpreter to
      embed.  Otherwise, it is searched in standard paths.

      In case you change the Python interpreter to embed, make sure that the
      Python scripts you use are written in the expected syntax for that
      version of the Python interpreter.  Notably, buffer objects have been
      replaced with memoryview objects in Python 3, and UTF-8 encoding now
      really matters for string literals (Python 3 uses bytes and Unicode
      objects).

      INN documentation and samples of Python hooks have been updated to
      provide more examples.

    * When a Python or Perl filter hook rejects an article, innd now
      mentions the reason in response to CHECK and TAKETHIS commands. 
      Previously, the reason was given only for the IHAVE command.

    * nnrpd now properly logs the hostname of clients whose connection
      failed owing to an issue during the negotiation of a TLS session or
      high load average.

Changes in 2.6.2

    * A new *syntaxchecks* parameter has been added in inn.conf.  It permits
      controlling the level of checks performed by innd and nnrpd.  Up to
      now, only one check can be enabled/disabled:  when *laxmid* is
      mentioned in the values of this new parameter, INN accepts Message-IDs
      that contain ".." in the left part, as well as Message-IDs with two
      "@" (such Message-IDs would otherwise be considered as syntactically
      invalid).  See the inn.conf(5) man page for more details.

      The check is disabled by default (*no-laxmid*), which corresponds to
      the legacy behaviour of INN 2.6.1 and earlier.

    * Use of the ovdb_server helper server is now the default when using the
      ovdb overview method, that is to say the default value for the
      *readserver* parameter in ovdb.conf is now set to true.  It improves
      stability and avoids deadlocks, timing issues and corrupted ovdb
      databases.

    * mailpost now removes empty header fields before attempting to post
      articles, and keeps trace of them in the X-Mailpost-Empty-Hdrs: newly
      generated header field body.  Also, mailpost now sanitizes header
      fields with regards to empty continuation header lines.  Thanks to
      Kamil Jonca for these bug reports.

    * A new -z parameter has been added to mailpost to mention a list of
      header fields to remove from the gated message.  Thanks to Dieter
      Stussy for the patch.

    * Fixed a bug in inews that was rejecting articles containing header
      fields whose length exceeded 998 bytes.  This limitation is for the
      length of a single line of a header field (and not for the length of
      the whole header field, as it was wrongly the case).

    * Added support for GnuPG's gpg binary (in addition to gpgv) in
      pgpverify.  Indeed, gpg still validates signatures made with weak
      digest algorithms like MD5 whereas gpgv no longer do.  Thanks to
      Thomas Hochstein for the patch, which permits validating control
      articles for hierarchies that are still using old PGP keys.

    * Added similar support for GnuPG's gpg binary in perl-nocem to validate
      NoCeM notices from issuers who are still using old PGP keys.

    * A few commands listed in the "Control commands to INND" section in
      daily Usenet reports were appearing as a mere letter; all of them are
      now properly converted to meaningful words.

    * The *tlsprotocols* parameter in inn.conf now recognizes the "TLSv1.3"
      value (for OpenSSL versions implementing TLS 1.3, that is to say
      starting from OpenSSL 1.1.1).

    * The buffindexed overview method will now hopefully work properly on
      systems with a native page size larger than 16KB.

    * Other minor bug fixes and documentation improvements.

Changes in 2.6.1

    * nnrpd now uses -0000 as the time zone for Date: and Injection-Date:
      header fields it generates.  It was previously using +0000, wrongly
      systematically indicating a local time zone at Universal Time when
      *localtime* is set to false (which is the default) in readers.conf. 
      The +0000 time zone will now be used only if *localtime* is set to
      true and UTC is really the local time zone of the server.

    * Julien Elie has implemented in nnrpd the new COMPRESS command
      described in draft-murchison-nntp-compress that extends the NNTP
      protocol to allow a connection to be effectively and efficiently
      compressed.  News clients that also support that extension will be
      able to benefit from that bandwidth optimization and improvement in
      speed.  Moreover, using COMPRESS is more secure than TLS-level

frontends/getlist.c
/*  $Id: getlist.c 10014 2016-05-05 12:42:31Z iulius $
**
**  Send a LIST command to an NNTP server and print the results.
*/

#include "config.h"
#include "clibrary.h"
#include <errno.h>

#include "inn/innconf.h"
#include "inn/messages.h"
#include "inn/nntp.h"
#include "inn/qio.h"
#include "inn/vector.h"
#include "inn/libinn.h"
#include "inn/paths.h"

static const char usage[] = "\
Usage: getlist [-AR] [-h host] [-p port] [list [pattern [types]]]\n\
\n\
getlist obtains a list from an NNTP server and prints it out.  By default,\n\
the active file is retrieved, but any list that the NNTP server supports\n\
can be requested.  The pattern is passed to the remote server as a filter\n\
on the returned values.  If the list parameter is \"active\", a third\n\
parameter can be given, listing the acceptable group types.  Only groups of\n\
that type (y and m being the most common) are returned.\n";


/*
**  Print out an appropriate error message for a bad NNTP response code and
**  exit.
*/
static void
die_nntp_code(enum nntp_code code, const char *line)
{
    if (code == 0)
        die("unexpected server response: %s", line);
    else
        die("unexpected server response: %03d %s", code, line);
}


/*
**  Print out an appropriate error message for a bad NNTP status and exit.
*/
static void
die_nntp_status(enum nntp_status status)
{
    switch (status) {
    case NNTP_READ_LONG:
        die("line from server too long for 128KB buffer");
    case NNTP_READ_EOF:
        die("server closed connection unexpectedly");
    case NNTP_READ_TIMEOUT:
        die("read from server timed out");
    default:
        sysdie("cannot read response from server");
    }
}


/*
**  Send a command and pattern to the remote server and make sure we get an
**  appropriate response.
*/
static void
send_list(struct nntp *nntp, const char *command, const char *pattern)
{
    bool okay;
    enum nntp_status status;
    enum nntp_code code;
    char *line;

    if (pattern != NULL)
        okay = nntp_send_line(nntp, "LIST %s %s", command, pattern);
    else
        okay = nntp_send_line(nntp, "LIST %s", command);
    if (!okay)
        sysdie("cannot send LIST command to server");
    status = nntp_read_response(nntp, &code, &line);
    if (status != NNTP_READ_OK)
        die_nntp_status(status);
    if (code != NNTP_OK_LIST)
        die_nntp_code(code, line);
}


/*
**  Print out the results of a LIST command.  Used for generic list commands
**  or active commands without a type parameter.
*/
static void
print_list(struct nntp *nntp)
{
    enum nntp_status status;
    char *line;

    status = nntp_read_line(nntp, &line);
    while (status == NNTP_READ_OK) {
        if (strcmp(line, ".") == 0)
            break;
        printf("%s\n", line);
        status = nntp_read_line(nntp, &line);
    }
    if (status != NNTP_READ_OK)
        die_nntp_status(status);
}


/*
**  Print out the results of a LIST ACTIVE command, limited to particular
**  group types.
*/
static void
print_active(struct nntp *nntp, const char *types)
{
    enum nntp_status status;
    char *line;
    struct cvector *group = NULL;

    status = nntp_read_line(nntp, &line);
    while (status == NNTP_READ_OK) {
        if (strcmp(line, ".") == 0)
            break;
        group = cvector_split_space(line, group);
        if (group->count != 4) {
            warn("malformed line from server: %s", line);
            continue;
        }
        if (strchr(types, group->strings[3][0]) != NULL)
            printf("%s %s %s %s\n", group->strings[0], group->strings[1],
                   group->strings[2], group->strings[3]);
        status = nntp_read_line(nntp, &line);
    }
    if (status != NNTP_READ_OK)
        die_nntp_status(status);
}


/*
**  Get the username and password for a remote site from the local password
**  file.  username and password will be set to newly allocated strings on
**  success.
*/
static bool
get_authinfo(const char *server, char **username, char **password)
{
    char *path, *line;
    QIOSTATE *passwords;
    struct cvector *info = NULL;

    path = concatpath(innconf->pathetc, INN_PATH_NNTPPASS);
    passwords = QIOopen(path);
    if (passwords == NULL) {
        if (errno != ENOENT)
            warn("cannot open %s", path);
        return false;
    }
    free(path);
    while ((line = QIOread(passwords)) != NULL) {
        if (line[0] == '\0' || line[0] == '#')
            continue;
        info = cvector_split(line, ':', info);
        if (info->count > 4 || info->count < 3)
            continue;
        if (info->count == 4 && strcmp(info->strings[3], "authinfo") != 0)
            continue;
        if (strcasecmp(info->strings[0], server) != 0)
            continue;
        *username = xstrdup(info->strings[1]);
        *password = xstrdup(info->strings[2]);
        return true;
    }
    return false;
}


/*
**  Send AUTHINFO information to a remote site.  Returns true if successful,
**  false on failure.  Problems sending to the remote site (as opposed to just
**  having the remote site reject the authentication) are fatal.
*/
static bool
send_authinfo(struct nntp *nntp, const char *username, const char *password)
{
    enum nntp_status status;
    enum nntp_code code;
    char *line;

    if (!nntp_send_line(nntp, "AUTHINFO USER %s", username))
        sysdie("cannot send AUTHINFO USER to remote server");
    status = nntp_read_response(nntp, &code, &line);
    if (status != NNTP_READ_OK)
        die_nntp_status(status);
    if (code == NNTP_OK_AUTHINFO)
        return true;
    if (code != NNTP_CONT_AUTHINFO)
        return false;
    if (!nntp_send_line(nntp, "AUTHINFO PASS %s", password))
        sysdie("cannot send AUTHINFO PASS to remote server");
    status = nntp_read_response(nntp, &code, &line);
    if (status != NNTP_READ_OK)
        die_nntp_status(status);
    return (code == NNTP_OK_AUTHINFO);
}


int
main(int argc, char *argv[])
{
    struct nntp *nntp;
    const char *host = NULL;
    const char *list = "active";
    const char *pattern = NULL;
    const char *types = NULL;
    enum nntp_status status;
    enum nntp_code response;
    char *line;
    unsigned short port = NNTP_PORT;
    bool authinfo = false;
    bool reader = false;
    int option;

    message_program_name = "getlist";
    if (!innconf_read(NULL))
        exit(1);
    host = innconf->server;

    /* Parse options. */
    while ((option = getopt(argc, argv, "Ah:p:R")) != EOF) {
        switch (option) {
        case 'A':
            authinfo = true;
            break;
        case 'h':
            host = optarg;
            break;
        case 'p':
            port = atoi(optarg);
            if (port == 0) {
                die("%s is not a valid port number", optarg);
            }
            break;
        case 'R':
            reader = true;
            break;
        default:
            die("%s", usage);
        }
    }
    argc -= optind;
    argv += optind;

    /* Read optional arguments. */
    if (argc > 3)
        die("too many arguments");
    if (argc >= 1)
        list = argv[0];
    if (argc >= 2)
        pattern = argv[1];
    if (argc == 3)
        types = argv[2];
    if (strcasecmp(list, "active") != 0 && types != NULL)
        die("group types can only be specified with a list type of active");

    /* Connect to the server. */
    if (host == NULL)
        sysdie("cannot get server name");
    nntp = nntp_connect(host, port, 128 * 1024, DEFAULT_TIMEOUT);
    if (nntp == NULL)
        sysdie("cannot connect to server %s:%hu", host, port);
    status = nntp_read_response(nntp, &response, &line);
    if (status != NNTP_READ_OK)
        die_nntp_status(status);
    if (response < 200 || response > 201)
        die_nntp_code(response, line);

    /* Switch to nnrpd if desired. */
    if (reader) {
        if (!nntp_send_line(nntp, "MODE READER"))
            sysdie("cannot send MODE READER command to server %s", host);
        status = nntp_read_response(nntp, &response, &line);
        if (status != NNTP_READ_OK)
            die_nntp_status(status);
    }

    /* Authenticate if desired. */
    if (authinfo) {
        char *username, *password;

        if (get_authinfo(host, &username, &password)) {
            if (!send_authinfo(nntp, username, password))
                warn("server %s did not accept authentication", host);
            free(username);
            free(password);
        } else
            warn("no authentication information found for %s", host);
    }

    /* Get and display the data. */
    send_list(nntp, list, pattern);
    if (types != NULL)
        print_active(nntp, types);
    else
        print_list(nntp);

    /* Be polite and say goodbye; it gives the server a chance to shut the
       connection down cleanly. */
    if (nntp_send_line(nntp, "QUIT"))
        nntp_read_response(nntp, &response, &line);
    nntp_free(nntp);
    exit(0);
}
